### ISSUE TITLE
Changing gems in build_config causes problems by leftover references

### ISSUE THREAD

**Issue by hlogmans** [mruby/mruby#3094]
When I change the included gems in build_config, I need to do a deep_clean. If I don't, the new gem is not included. If I removed a gem, part of the references seem to be still there and the C compiler crashes during 'make test'.

````
...
LD    build/test/bin/mrbtest 
Undefined symbols for architecture x86_64:
  "_GENERATED_TMP_mrb_ruby_extension_example_gem_test", referenced from:
      _mrbgemtest_init in mrbtest.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Undefined symbols for architecture x86_64:
  "_GENERATED_TMP_mrb_ruby_extension_example_gem_test", referenced from:
      _mrbgemtest_init in mrbtest.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
rake aborted!
Command Failed: [gcc  -o "/Users/hlogmans/development/mruby/usb2/build/test/bin/mrbtest" "/Users/hlogmans/development/mruby/usb2/build/test/mrbgems/mruby-test/driver.o" "/Users/hlogmans/development/mruby/usb2/build/test/mrbgems/mruby-test/mrbtest.o" "/Users/hlogmans/development/mruby/usb2/build/test/mrbgems/mruby-test/mrbtest.a" "/Users/hlogmans/development/mruby/usb2/build/test/lib/libmruby_core.a" "/Users/hlogmans/development/mruby/usb2/build/test/lib/libmruby.a"  -lm -lreadline -lncurses ]

`````

If I run `rake  deep_clean` and then build, it runs fine. This is nowhere documents and by accident I noticed the deep_clean command...

Using lastest mruby from git.

**Comment by matz** [comment]
@zzak is working on simplifying build process. I hope this will be addressed in the future. Until then, try `rake clean` after you modified `build_config.rb`.

**Comment by zzak** [comment]
Yes, I have experimental build system that I've been hacking on under [zzak/mruby-build](https://github.com/zzak/mruby-build).

I think the problem you mention has to deal with the `mrbtest` binary, and probably some missing file dependency in rake. I'll see if I can figure it out!

**Comment by bovi** [comment]
@hlogmans could you list out the steps you have performed to reach this error message? I certainly believe that there is some dirt effects during the compilation process but I could not reproduce it.

**Comment by bovi** [comment]
I could reproduce the exact error when performing the following steps:
  - do a clean build with the standard build_config.rb
  - remove ```conf.gembox 'default'```
  - attempt to build again without ```rake clean```

The problem here was that mrbtest.c in the mruby-test mgem has no prerequisite to update itself after it was created. This should be fixed by #3101

**Comment by hlogmans** [comment]
@bovi Great! I will check it out ASAP. I'm only using the mrbtest functionality as I'm building a mRuby-C test system that helps in easily building unit tests for C functions and libraries. And the software to be tested is provided as a gem with unit tests. 

When checking I will also check some other cases as I found that there seem to be some other issues with leftovers in the build folder. A rake clean always did the job.
